Test: add rhsmClient status helper for associated template tests

swadeley opened the pull request:
## Summary

Add a helper that checks if an rhsmClient is present in inventory and attached to a template
Make use of that helper in Associated Template test

## Testing steps
Integration/AssociatedTemplateCRUD.spec.ts passes

swadeley committed:
Remove unnecessary navigation steps

   where the test is already on the Templates page

swadeley committed:
Add system-to-template attachment helper

  Polls to check if system is attached to a template.

swadeley committed:
Add host name helper to rhsmClient

  This gets the display name of the container.
  The name is generated by the hostname rpm in the container.

swadeley committed:
Poll for host-to-template status

  Poll from test to confirm host-to-template attachement status before UI check.

marusak commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -74,9 +77,15 @@ test.describe('Associated Template CRUD', async () => {
 
+    await test.step('Verify system is attached to template', async () => {
+      hostname = await regClient.GetHostname();
+      console.log('System hostname:', hostname);
```
Is this debugging or we want to keep it in? :thinking:

marusak commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -74,9 +77,15 @@ test.describe('Associated Template CRUD', async () => {
+      hostname = await regClient.GetHostname();
+      console.log('System hostname:', hostname);
+      const isAttached = await pollForSystemTemplateAttachment(page, hostname, 10_000, 6);
+      expect(isAttached).toBe(true);
```
I just learned the other day that "expect" can have second argument. See https://playwright.dev/docs/test-assertions#custom-expect-message

That is cool, because having these bare "toBe(true)" show up in logs just as "toBe(True)" which does not say a lot. This step is very simple, it has one assert so it would be straight forward what failed, but may be a good practice to start adding messages to these asserts

swadeley replied:
Ok, sounds good, I will do that when other changes have been made.

swadeley committed:
Handle "already unregistered" in client cleanup

swadeley commented on _playwright-tests/Integration/helpers/systemHelpers.ts:
```diff
@@ -0,0 +1,92 @@
+      if (!system) {
+        // System not found in inventory; it should not be attached to a template
+        console.log(`System '${hostname}' not found in inventory (not attached)`);
+        return false;
```
Ok, good point, will change to poll and log attempts same as the other steps.

swadeley commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -74,9 +77,15 @@ test.describe('Associated Template CRUD', async () => {
+
     await test.step('Attempt to delete template and verify warning appears', async () => {
       await navigateToTemplates(page);
-
```
Hi, there is a commit to fix that in older PR: https://github.com/content-services/content-sources-frontend/pull/696/commits/2cf8fe62852b990ffe3b449a21978705f7aff5e3

swadeley commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -107,6 +116,13 @@ test.describe('Associated Template CRUD', async () => {
+    await test.step('Wait for system to be removed from inventory', async () => {
+      // Poll until system is not found or not attached to template
+      // This should ensure the backend has processed the unregistration before checking UI
+      const isAttached = await pollForSystemTemplateAttachment(page, hostname, 10_000, 6);
```
I will wait for you further review.

swadeley committed:
Add custom expect messages

  PW custom expect messages provide more context for assertions.

katarinazaprazna commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -104,38 +122,56 @@ test.describe('Associated Template CRUD', async () => {
         await expect(
           modal.getByRole('link', { name: /This template is assigned to \d+ system/i }),
+          'system assignment link should not be visible',
         ).not.toBeVisible();
```
I've recently learned that the direct opposite of `.toBeVisible()` is `.toBeHidden()`. We could try using `.toBeHidden()` as it is more descriptive of the intent in the code

swadeley replied:
Use negative matcher where possible #733

katarinazaprazna commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -104,38 +122,56 @@ test.describe('Associated Template CRUD', async () => {
+        await expect(
+          rowTemplate.getByText('Valid'),
+          'template should not be visible in the list',
+        ).not.toBeVisible({ timeout: 30000 });
```
Could we apply the same suggestion as mentioned in my comment above regarding the `.toBeHidden()` matcher?

swadeley replied:
Fixed that, thank you

swadeley replied:
I decided to use `toHaveCount(0)`

katarinazaprazna reviewed (changes requested):
I've completed my review and left a few comments outlining some minor suggestions. Overall, the changes look great! Thanks for putting this together :)

@marusak, would you be able to provide a quick re-review? Thanks so much!

swadeley commented on _playwright-tests/Integration/helpers/rhsmClient.ts:
```diff
@@ -193,16 +205,37 @@ export class RHSMClient {
   async Destroy(unregisterMethod: 'rhc' | 'sm' | 'none' = 'none') {
```
Ok, changes made. Thank you.

marusak commented on _playwright-tests/Integration/AssociatedTemplateCRUD.spec.ts:
```diff
@@ -107,6 +116,18 @@ test.describe('Associated Template CRUD', () => {
+      expect(isAttached).toBe(true);
+    });
+
+    await test.step('Reload page to ensure UI reflects system removal', async () => {
```
Ugh, do we really need to reload the page?

marusak commented:
Looking good. One note, one question (or maybe I am just confused about usage of .not.toBeEnabled()). Mainly though you added page reloading. That is a big anti-pattern. The whole point about not using `NavigateTo...()` multiple times was to avoid reloading pages. Do we have to do it as the page never updates? If so, then that is a bug. (We can report it and add note why we are doing this workaround)

swadeley committed:
Reload page to force system list update

  Revert this commit when HMS-9731 is resolved
  https://issues.redhat.com/browse/HMS-9731

marusak reviewed (approved):
Thank you!

katarinazaprazna reviewed (approved):
Thank you so much for taking the time to make these thoughtful changes! ✅
